### PR TITLE
Add (Mini/Mamba)forge

### PR DESCRIPTION
**Pull request by bollwyvl**
## Related
- one approach to #76

## Code Changes
- `download-miniforge.ts`
  - uses GitHub API to fetch releases (if version is not specified)... am trying to properly cache these, but i think there's some cache locality stuff
 - removes some dead code
 - tweaks provider logging to avoid constructions like `Can we use use conda create?`
- `conda.ts`
 - if `use-mamba` ~~and `mamba-in-installer`~~ is specified, will try to use `mamba` earlier in the process, but also catches things like `conda init` which _aren't_ supported by the `mamba` CLI yet
- `installer/index.ts`
  - now accepts (and can update) dynamic options, so that it can detect an installed `mamba`  
- `base-tools/index.ts`
  - splits up tool specs and post-install
- `upgrade-mamba.ts`
  - can no-up the upgrading, but we will _still_ want the mamba bash wrapper on windows without starting the solver)
- tests
  - adds workflow with a shotgun of different options to test out versions, interplay with other stuff 

## User-facing changes
- adds `miniforge-variant`, `miniforge-version`, to specify a Miniforge product
- adds ~~`mamba-in-installer`~~ and `use-mamba` to fine-tune use of Mambaforge
- updates the README, with some cross-links between mamba, custom installer, and this new provider

## Backwards-incompatible changes
N/A

## Follow-on

- with these changes, we're ready for `micromamba` (#75), but: 
  - the `mamba` CLI doesn't yet appear to support `init` for shell integration... 
    - perhaps we can hot-install `micromamba` into the bundled miniconda `bin` or `Scripts` or `condabin` or something?
- while our list of inputs (and therefore our README) is getting rather long, I think from a maintenance perspective, keeping _one_ action running that can do all the things is probably useful vs splitting effort (especially now that we've #107-ed ourselves to where we are now)
  - perhaps we should consider doing a real docs site, potentially with a little "wizard" for building up a workflow
  - or, start using the GitHub wiki or discussions

### CODE REVIEW

**Comment by bollwyvl** [comment]
Hooray :green_heart: !

**Comment by goanpeca** [comment]
Thanks for working on this @bollwyvl, this is awesome :)

File: action.yml [hunk 1]
```diff
@@ -180,13 +200,26 @@ inputs:
       removed from the runner. Default is "true".'
     required: false
     default: "true"
+  mamba-in-installer:
```
Comment (jaimergp) [inline action.yml:203]: Can this be checked programmatically? Once installed one can check for `mamba` or `mamba.bat` in `$PREFIX/condabin` and determine it without user intervention, I think.

Comment (bollwyvl) [reply action.yml:203]: hmmm... it's probably worth a look...

**Comment by goanpeca** [comment]
>perhaps we should consider doing a real docs site, potentially with a little "wizard" for building up a workflow

Absolutely, I think we are big enough that we can have dedicated docs :) will set something with githubpages and myst :)

**Comment by bollwyvl** [comment]
Ok, so ab60b01 is back to :green_circle: and 4651cc0 is mostly cosmetic... thanks @jaimergp for pushing to remove the manual thing: doing the detection uncovered a number of future bugs (e.g. adding that bash wrapper on windows).

`example-10.yml` is getting pretty hairy at this point, but I think has proven out a lot of cases (e.g. mamba in the installer upgrading to a specific mamba). I'd better stop tinkering with it.

Let's _definitely_ squash merge this one :P

**Comment by jaimergp** [comment]
Nice!

>  adding that bash wrapper on windows

Hopefully this is addressed at the mamba level at some point (them providing a shell command like conda does)

**Comment by bollwyvl** [comment]
> Nice!
>
> adding that bash wrapper on windows
>
> Hopefully this is addressed at the mamba level at some point (them
> providing a shell command like conda does)
>
Indeed. Guess we should issue? I'm a little mamba-ed out right now.

God help zsh/fish windows users... But I could see xonsh, frankly...
P-strings are awesome for windows "specialness".

**Comment by goanpeca** [comment]
Another awesome piece of work @bollwyvl :) 🥳 

Thanks!

**Pull request merged**